stan/math/prim/fun/col.hpp

```diff
@@ -14,17 +14,16 @@ namespace math {
  * This is equivalent to calling <code>m.col(i - 1)</code> and
  * assigning the resulting template expression to a column vector.
  *
- * @tparam T type of elements in the matrix
+ * @tparam T type of the matrix
  * @param m Matrix.
  * @param j Column index (count from 1).
  * @return Specified column of the matrix.
  * @throw std::out_of_range if j is out of range.
  */
-template <typename T>
-inline Eigen::Matrix<T, Eigen::Dynamic, 1> col(
-    const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& m, size_t j) {
+template <typename T, typename = require_eigen_t<T>>
+inline auto col(const T& m, size_t j) {
   check_column_index("col", "j", m, j);
-  return m.col(j - 1);
+  return m.col(j - 1).eval();
```

Comment:
Please add a "xyz_works_with_other_functions" test for this function so someone doesnt remove the eval for some reason before we start accepting eigen expressions.

Reply:
Which other function should I test this with? In #1644 I used multiply, but it has been generalized in #1628, so these tests won't fail anymore if eval is removed. The problem is that regardless of what function we choose such tests won't detect missing eval anymore once the function is generalized. I could make one non-generalized function just for testing, but then we will have to remove many tests once all functions accept expressions.

Reply:
Yeah, that is a fair point. Whichever function we use to test with is going to not be a valuable test once its generalized. Well I guess there isnt a good solution here.

Reply:
And we need to test this on the upstream levels, which we said before. Will resolve these then.